File: packages/capacitor-plugin/android/src/main/java/com/otakit/updater/ForegroundDeadline.java
package com.otakit.updater;

import android.os.Handler;
import android.os.SystemClock;
import java.util.function.LongSupplier;

/** Owned by the main thread. Only foreground time consumes the readiness budget. */
final class ForegroundDeadline {

  interface Scheduler {
    Runnable schedule(long delayMs, Runnable callback);
  }

  private final LongSupplier now;
  private final Scheduler scheduler;
  private boolean foreground;
  private long remainingMs;
  private long startedAt;
  private long generation;
  private Runnable action;
  private Runnable cancelScheduled;

  ForegroundDeadline(Handler handler) {
    this(SystemClock::uptimeMillis, (delay, callback) -> {
      handler.postDelayed(callback, delay);
      return () -> handler.removeCallbacks(callback);
    });
  }

  ForegroundDeadline(LongSupplier now, Scheduler scheduler) {
    this.now = now;
    this.scheduler = scheduler;
  }

  void start(long timeoutMs, Runnable action) {
    cancel();
    remainingMs = Math.max(0, timeoutMs);
    this.action = action;
    arm();
  }

  void setForeground(boolean value) {
    if (foreground == value) return;
    if (foreground && action != null) {
      remainingMs = Math.max(0, remainingMs - Math.max(0, now.getAsLong() - startedAt));
    }
    disarm();
    foreground = value;
    arm();
  }

  void cancel() {
    disarm();
    action = null;
    remainingMs = 0;
  }

  private void disarm() {
    generation++;
    if (cancelScheduled != null) cancelScheduled.run();
    cancelScheduled = null;
  }

  private void arm() {
    if (!foreground || action == null) return;
    startedAt = now.getAsLong();
    long expectedGeneration = ++generation;
    cancelScheduled = scheduler.schedule(remainingMs, () -> fire(expectedGeneration));
  }

  private void fire(long expectedGeneration) {
    if (generation != expectedGeneration || !foreground || action == null) return;
    remainingMs = Math.max(0, remainingMs - Math.max(0, now.getAsLong() - startedAt));
    cancelScheduled = null;
    if (remainingMs > 0) {
      arm();
      return;
    }
    Runnable expired = action;
    action = null;
    generation++;
    expired.run();
  }
}

File: packages/capacitor-plugin/ios/Sources/UpdaterPlugin/ForegroundDeadline.swift
import Foundation

/// Owned by the main thread. Only foreground time consumes the readiness budget.
final class ForegroundDeadline {
  typealias Scheduler = (TimeInterval, @escaping () -> Void) -> () -> Void
  private let now: () -> TimeInterval
  private let schedule: Scheduler
  private var foreground = false
  private var remaining: TimeInterval = 0
  private var startedAt: TimeInterval = 0
  private var generation = 0
  private var action: (() -> Void)?
  private var cancelScheduled: (() -> Void)?
  private var observations: [(NotificationCenter, NSObjectProtocol)] = []

  init(
    now: @escaping () -> TimeInterval = { ProcessInfo.processInfo.systemUptime },
    schedule: @escaping Scheduler = { delay, callback in
      let item = DispatchWorkItem(block: callback)
      DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: item)
      return { item.cancel() }
    }
  ) {
    self.now = now
    self.schedule = schedule
  }

  func observe(
    center: NotificationCenter, active: Notification.Name, inactive: Notification.Name,
    initiallyActive: Bool
  ) {
    setForeground(initiallyActive)
    observations.append((center, center.addObserver(forName: active, object: nil, queue: .main) { [weak self] _ in
      self?.setForeground(true)
    }))
    observations.append((center, center.addObserver(forName: inactive, object: nil, queue: .main) { [weak self] _ in
      self?.setForeground(false)
    }))
  }

  func start(timeout: TimeInterval, action: @escaping () -> Void) {
    cancel()
    remaining = max(0, timeout)
    self.action = action
    arm()
  }

  func setForeground(_ value: Bool) {
    guard value != foreground else { return }
    if foreground, action != nil {
      remaining = max(0, remaining - max(0, now() - startedAt))
    }
    disarm()
    foreground = value
    arm()
  }

  func cancel() {
    disarm()
    action = nil
    remaining = 0
  }

  private func disarm() {
    generation += 1
    cancelScheduled?()
    cancelScheduled = nil
  }

  private func arm() {
    guard foreground, action != nil else { return }
    startedAt = now()
    generation += 1
    let expectedGeneration = generation
    cancelScheduled = schedule(remaining) { [weak self] in
      self?.fire(expectedGeneration: expectedGeneration)
    }
  }

  private func fire(expectedGeneration: Int) {
    guard generation == expectedGeneration, foreground, let action else { return }
    remaining = max(0, remaining - max(0, now() - startedAt))
    cancelScheduled = nil
    if remaining > 0 { arm(); return }
    self.action = nil
    generation += 1
    action()
  }

  deinit {
    cancelScheduled?()
    for (center, observation) in observations { center.removeObserver(observation) }
  }
}
